src/ray/raylet/worker_pool.cc

```diff
@@ -152,8 +152,19 @@ void WorkerPool::SetAgentManager(std::shared_ptr<AgentManager> agent_manager) {
   agent_manager_ = agent_manager;
 }
 
+void WorkerPool::PopWorkerCallbackExecution(const PopWorkerCallback &callback,
+                                            std::shared_ptr<WorkerInterface> worker,
+                                            Status status) {
+  if (callback) {
+    // Call back this function asynchronously if needed. Such as
+    // io_service_->post([callback, worker, status]() { callback(worker, status); });
+    callback(worker, status);
```

Comment:
I think we should always post it instead of calling in the same stack. This can cause many unexpected weird issues

Reply:
I agree that using post is more reasonable. But I found both `worker_pool_test` and `cluster_task_manager_test` are synchronous test. I should refactor tests any more.

Reply:
I remember there were cases like https://github.com/ray-project/ray/issues/2959. I guess it is not happening in the current code structure, but there's nothing to lose by being more careful